src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/StringCache.cs

```diff
@@ -0,0 +1,158 @@
+
+        public string GetOrAddValue(string key)
+        {
+            lock (_lock)
+            {
+                if (key is null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                if (TryGetValue(key, out var result))
+                {
+                    return result!;
+                }
+                else
+                {
+                    var weakRef = new Entry(key);
+                    _hashSet.Add(weakRef);
+
+                    // Whenever we expand lets clean up dead references
+                    if (_capacity <= _hashSet.Count)
+                    {
+                        Cleanup();
+                        _capacity *= 2;
+                    }
+
+                    return key;
+                }
+            }
+        }
+
+        private void Cleanup()
+        {
+            _hashSet.RemoveWhere((weakRef) =>
+            {
+                return !weakRef.IsAlive;
+            });
+        }
+
+        private bool TryGetValue(string key, out string? value)
+        {
+            if (_hashSet.Contains(new Entry(key)))
+            {
+                var exfiltrator = (ExfiltratingEqualityComparer)_hashSet.Comparer;
+                var val = exfiltrator.LastEqualValue!;
+                if (val.TryGetTarget(out var target))
+                {
+                    value = target!;
+                    return true;
+                }
+                else
+                {
+                    // If one of our strings is out of scope probably tons of them are.
+                    Cleanup();
+                }
+            }
+
+            value = default;
+            return false;
+        }
+
+        /// <summary>
+        /// This is a gross hack to do a sneaky and get the value inside the HashSet out given the lack of any Get operations in netstandard2.0.
+        /// If we ever upgrade to 2.1 delete this and just use the built in.
+        /// </summary>
+        /// <remarks>
+        /// This is fragile on the ordering of the values passed to the EqualityComparer by HashSet.
+        /// If that ever switches we have to react, if it becomes indeterminate we have to abandon this strategy.
+        /// </remarks>
+        private class ExfiltratingEqualityComparer : IEqualityComparer<Entry>
+        {
+            public Entry? LastEqualValue { get; private set; }
+
+            public bool Equals(Entry x, Entry y)
+            {
+                if (x.Equals(y))
+                {
+                    LastEqualValue = x;
+                    return true;
+                }
+                else
+                {
+                    LastEqualValue = null;
+                    return false;
+                }
+            }
+
+            public int GetHashCode(Entry obj)
+            {
+                return obj.TargetHashCode;
+            }
+        }
+
+        private class Entry
```

Comment:
Why even have an entry if you're comparing the value/hashcode anyhow? Aka couldn't you just have a Dictionary<int, WeakReference<string>>?

Reply:
`Dictionary<int, WeakReference>` can't handle Hash collisions, and especially if we start using this in other places I think that's a real possibility.

Reply:
I think I'm missing something, how does what you have now handle it better?

Reply:
There was a concern before I responded to your comment about using hash in Equal as a fallback, but as far as I know what I currently have should be resilient to hash collisions due to normal HashSet bucketing.

And WRT the concerns about using an `Entry` class here, that's the same basic idea as the internals of `ConditionalWeakReferenceTable`.

If that still doesn't answer you lets jump on a call.

Reply:
Talked offline. Makes sense.